rasterio/dtypes.py

```diff
@@ -203,16 +203,11 @@ def can_cast_dtype(values, dtype):
     boolean
         True if values can be cast to data type.
     """
-    values = np.asarray(values)
-
-    if values.dtype.name == _getnpdtype(dtype).name:
+    dtype_name = _getnpdtype(dtype).name
+    if is_ndarray(values) and values.dtype.name == dtype_name:
         return True
-
-    elif values.dtype.kind == 'f':
-        return np.allclose(values, values.astype(dtype), equal_nan=True)
-
-    else:
-        return np.array_equal(values, values.astype(dtype))
+    min_dtype = get_minimum_dtype(values)
+    return np.can_cast(min_dtype, dtype_name, casting='safe')
```

Comment:
I see in https://numpy.org/doc/stable/reference/generated/numpy.can_cast.html, that it can accept an array:
```suggestion
    return np.can_cast(values, dtype_name, casting='safe')
```
Have you tried this out to see what the performance is like?

Reply:
I didn't do this because it won't allow you to downcast (ie go from 'float64' to 'float32' if all your data fits in float32).